### DISCUSSION TITLE
Debugging Federated Learning Pipeline

### DISCUSSION THREAD

**Discussion by LSnyd** [NVIDIA/NVFlare#281] [Q&A] [answered]
Hi all, 
does anyone by chance have any tips on how to debug the setup of a federated learning pipeline? 

I am having quite some issues tracing the errors that I am getting for instance when adding Homomorphic Encryption to my pipeline. From the error messages, it is not always clear for me which specific configuration is causing the error. 

I'd be super happy for any hints. Thank you!

**Comment by yanchengnv** [comment, marked as answer]
Thanks for the question.

Debugging a distributed and componentized system is quite challenging in general. By "componentized", I mean that the system function is achieved by a collection of configured components (Python objects).

That said, we usually follow these steps:

- The key is to determine which component(s) caused the error. From the error message (or exception traceback), try to determine which component is complaining. 
- Components are configured in various config JSON files (config_fed_server.json, config_fed_client.json in your app definition). These objects are created at the time of system sttartup.
- For errors on Server, check config_fed_server.json; for errors on Client, check config_fed_client.json.

Once the component is found, check its configuration or code for the bug.

Hope this helps.

**Reply by LSnyd** [reply]
Hi @yanchengnv, 
I am currently taking the same approach to debug as you described, so seems like I am already doing the right thing.  
Thanks a lot for your explanation though!
